### PR TITLE
DEV-634 workaround for proxy requests with boto

### PR DESCRIPTION
**Pull request by ProfOak**

### CODE REVIEW

File: cdisutils/storage.py [hunk 1]
```diff
@@ -8,25 +8,118 @@
...
+        if resp.status != 200:
+            # Fake a socket error, use a code that make it obvious it hasn't
+            # been generated by the socket library
+            raise socket.error(-71,
+                               "Error talking to HTTP proxy %s:%s: %s (%s)" %
+                               (self.proxy, self.proxy_port,
+                                resp.status, resp.reason))
+
+        # We can safely close the response, it duped the original socket
+        resp.close()
```
Comment (BenX00) [inline cdisutils/storage.py:88]: I believe we are closing a file pointer with this. We should make sure to put the above logic into a try block and this close into a finally block.

File: cdisutils/storage.py [hunk 2]
```diff
@@ -8,25 +8,118 @@
...
+    Amazon S3 because it needs a proxy to access it.
+
+    With the Python 3 upgrade came breaking changes specifically pertaining to
+    bytes and strings. The sendall() socket command no longer accepts strings.
+    We have to convert them all to bytes before sending.
+
+    There have been attempts to merge a fix for this into the main boto branch
+    but they have not been successful.
+    - https://github.com/boto/boto/pull/2718
+    - https://github.com/boto/boto/pull/3699
```
[the remark this answers is not in the document]
Comment (BenX00) [reply cdisutils/storage.py:45]: :tada:

File: cdisutils/storage.py [hunk 3]
```diff
@@ -8,23 +8,130 @@
...
+    """
+    def __init__(self, *args, **kwargs):
+        super(S3ConnectionProxyFix, self).__init__(*args, **kwargs)
+        self.logger = cdisutils.log.get_logger("S3ConnectionProxyFix")
+
+    def _get_host(self, host=None, port=None):
+        if host and port:
+            return '%s:%d' % (host, port)
+        else:
+            return '%s:%d' % (self.host, self.port)
```
Comment (mpsolano) [inline cdisutils/storage.py:55]: Just a minor style suggestion to use double-quotes here and in a couple other places in this class.

File: cdisutils/storage.py [hunk 4]
```diff
@@ -8,23 +8,130 @@
...
+                # Fake a socket error, use a code that make it obvious it hasn't
+                # been generated by the socket library
+                raise socket.error(
+                    -71,
+                    "Error talking to HTTP proxy %s:%s: %s (%s)" %
+                    (self.proxy, self.proxy_port, resp.status, resp.reason),
+                )
+
+        except Exception as e:
+            print(e)
```
Comment (mpsolano) [inline cdisutils/storage.py:92]: Would it make sense to replace the print statement with a log statement?

File: cdisutils/storage.py [hunk 5]
```diff
@@ -8,23 +8,130 @@
 TODO: add tests
```
Comment (mpsolano) [inline cdisutils/storage.py:8]: Do we still have outstanding tests to write?

**Review by mpsolano — APPROVED** [review]
Looks good!

File: cdisutils/storage.py (hunk 5 above)
Comment (ProfOak) [reply cdisutils/storage.py:8]: That's possible but but I'm not sure. This comment was added 6 years ago by someone who no longer works here. The existing tests were changed to work with the `S3ConnectionProxyFix` class.

File: cdisutils/storage.py [hunk 6]
```diff
@@ -8,23 +8,130 @@
...
+import json
+import re
+import socket
+import ssl
 
-from .log import get_logger
-from boto import connect_s3
+from boto.compat import http_client
+from boto import config, UserAgent, https_connection
+from boto.s3.connection import S3Connection
```
Comment (BenX00) [inline cdisutils/storage.py:19]: You should import the namespaces and not the objects as per the style guide.

**Review by BenX00 — APPROVED** [review]
Just saw little things.

**Pull request merged**